### PR TITLE
crypto/merkle: pre-allocate data slice in innherHash

### LINKED ISSUE #6442: crypto/merkle: innherHash use more memory than necessary
## Summary

Speed up `innerHash`

## Problem Definition

Benchmarking on `HashAlternatives` shows a bottle neck at `innerHash`:

```
         .          .     23:// returns tmhash(0x01 || left || right)
         .          .     24:func innerHash(left []byte, right []byte) []byte {
    7.73GB     9.42GB     25:	return tmhash.Sum(append(innerPrefix, append(left, right...)...))
         .          .     26:}
```

The `append(innerPrefix, append(left, right...)...)` operation is clearly the culprit, since when the runtime must check for slice capacity before appending to it.

## Proposal

We can pre-allocate the data slice, which have enough capacity to hold all the data. That will reduce the memory usage, and also speed `HashAlternatives` a bit.

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by cuonglm**
So we can reduce pressure on runtime for checking that slice has enough
capacity before appending.

Benchmark result shows less memory usage and faster runtime:

```
name                          old time/op    new time/op    delta
HashAlternatives/recursive-8    25.2µs ± 1%    19.6µs ± 1%  -22.07%  (p=0.000 n=10+9)
HashAlternatives/iterative-8    24.2µs ± 1%    19.0µs ± 0%  -21.28%  (p=0.000 n=10+10)

name                          old alloc/op   new alloc/op   delta
HashAlternatives/recursive-8    25.4kB ± 0%    19.1kB ± 0%  -24.92%  (p=0.000 n=10+10)
HashAlternatives/iterative-8    28.1kB ± 0%    21.8kB ± 0%  -22.54%  (p=0.000 n=10+10)

name                          old allocs/op  new allocs/op  delta
HashAlternatives/recursive-8       497 ± 0%       398 ± 0%  -19.92%  (p=0.000 n=10+10)
HashAlternatives/iterative-8       498 ± 0%       399 ± 0%  -19.88%  (p=0.000 n=10+10)
```

Fixes #6442

### CODE REVIEW

**Comment by cuonglm** [comment]
cc @odeke-em

**Review by marbar3778 — APPROVED** [review]
Could you add a changelog entry as well please? We like to give credit to contributors.

**Review by odeke-em — APPROVED** [review]
LGTM, thank you @cuonglm!

**Comment by odeke-em** [comment]
One other thing we could do much later on after this PR is to add a sync.Pool to reuse slices since the byte slices are passed into tmhash.Sum and read-only.

File: crypto/merkle/hash.go [hunk 1]
```diff
@@ -22,5 +22,9 @@ func leafHash(leaf []byte) []byte {
 
 // returns tmhash(0x01 || left || right)
 func innerHash(left []byte, right []byte) []byte {
-	return tmhash.Sum(append(innerPrefix, append(left, right...)...))
+	data := make([]byte, 0, len(innerPrefix)+len(left)+len(right))
+	data = append(data, innerPrefix...)
+	data = append(data, left...)
+	data = append(data, right...)
```
Comment (odeke-em) [inline crypto/merkle/hash.go:28]: How about using copy?
```go
n := copy(data, innerPrefix)
n += copy(data[n:], left)
n += copy(data[n:], right)
```
Please try that and run some benchmarks then let me know what's up.

Comment (odeke-em) [reply crypto/merkle/hash.go:28]: Awesome, let's use it then! I believe that copy translates into a MEMMOVE which is much much faster and as the bytes to be hashed increased we should even more radical improvements.

Comment (odeke-em) [reply crypto/merkle/hash.go:28]: Also please don't forget to use `tmhash.Sum(data[:n])` :-)

File: crypto/merkle/hash.go [hunk 2]
```diff
@@ -22,5 +22,9 @@ func leafHash(leaf []byte) []byte {
 
 // returns tmhash(0x01 || left || right)
 func innerHash(left []byte, right []byte) []byte {
-	return tmhash.Sum(append(innerPrefix, append(left, right...)...))
+	data := make([]byte, 0, len(innerPrefix)+len(left)+len(right))
+	n := copy(data, innerPrefix)
+	n = copy(data[n:], left)
+	copy(data[n:], right)
+	return tmhash.Sum(data)
```
Comment (odeke-em) [inline crypto/merkle/hash.go:29]:
```go
n += copy(data[n:], right)
return tmhash.Sum(data[:n])
```

File: crypto/merkle/hash.go [hunk 3]
```diff
@@ -22,5 +22,9 @@ func leafHash(leaf []byte) []byte {
 
 // returns tmhash(0x01 || left || right)
 func innerHash(left []byte, right []byte) []byte {
-	return tmhash.Sum(append(innerPrefix, append(left, right...)...))
+	data := make([]byte, 0, len(innerPrefix)+len(left)+len(right))
+	n := copy(data, innerPrefix)
+	n = copy(data[n:], left)
```
Comment (odeke-em) [inline crypto/merkle/hash.go:27]: Oh oh, this is invalid, it has to be a: `n += copy(data[n:], left)` otherwise you'll just have the length of left

Comment (cuonglm) [reply crypto/merkle/hash.go:27]: Yay, we need to pre-allocate all slice elem 😄

**Comment by odeke-em** [comment]
@marbar3778 @tychoish please help us merge lest the PR gets stale. Thank you.

**Pull request merged**